File: node/AssemblerTest/preprocessExtensions.js
// Extension methods for PreprocessedSiteTemplates to match C# structure
// This file corresponds to PreprocessExtensions.cs in the C# implementation

/**
 * Extension methods for PreprocessedSiteTemplates
 * These methods provide serialization and summary functionality for preprocessed templates
 */

/**
 * Serializes the preprocessed site templates to a formatted JSON string for debugging and validation
 * @param {PreprocessedSiteTemplates} siteTemplates - The preprocessed site templates
 * @param {boolean} indented - Whether to format the JSON with indentation for readability
 * @returns {string} JSON representation of the preprocessed templates structure
 */
export function toJson(siteTemplates, indented = true) {
    return serializePreprocessedTemplates(siteTemplates, indented);
}

/**
 * Creates a summary object containing key statistics about the preprocessed templates
 * @param {PreprocessedSiteTemplates} siteTemplates - The preprocessed site templates
 * @returns {Object} Summary object with template statistics
 */
export function createSummary(siteTemplates) {
    let templatesRequiringProcessing = 0;
    let templatesWithJsonData = 0;
    let templatesWithPlaceholders = 0;

    for (const template of siteTemplates.templates.values()) {
        if (template.requiresProcessing) {
            templatesRequiringProcessing++;
        }
        if (template.hasJsonData) {
            templatesWithJsonData++;
        }
        if (template.hasPlaceholders) {
            templatesWithPlaceholders++;
        }
    }

    return {
        siteName: siteTemplates.siteName,
        templatesRequiringProcessing,
        templatesWithJsonData,
        templatesWithPlaceholders,
        totalTemplates: siteTemplates.templates.size
    };
}

/**
 * Serializes the summary to JSON for easier analysis
 * @param {PreprocessedSiteTemplates} siteTemplates - The preprocessed site templates
 * @param {boolean} indented - Whether to format the JSON with indentation
 * @returns {string} JSON representation of the template summary
 */
export function toSummaryJson(siteTemplates, indented = true) {
    return serializePreprocessedSummary(createSummary(siteTemplates), indented);
}

// Serialization helpers (formerly in JsonConverter)
/**
 * Serializes preprocessed templates to JSON format
 * @param {PreprocessedSiteTemplates} templates - The preprocessed templates to serialize
 * @param {boolean} indented - Whether to indent the JSON output
 * @returns {string} JSON string representation
 */
export function serializePreprocessedTemplates(templates, indented = true) {
    // Sort templates to match C# ordering (main template first, then content templates)
    const sortedEntries = Array.from(templates.templates.entries()).sort(([keyA], [keyB]) => {
        // Main template (without "content" suffix) comes first
        const isMainA = !keyA.includes('content');
        const isMainB = !keyB.includes('content');
        if (isMainA && !isMainB) return -1;
        if (!isMainA && isMainB) return 1;
        return keyA.localeCompare(keyB);
    });

    const obj = {
        siteName: templates.siteName,
        templates: Object.fromEntries(
            sortedEntries.map(([key, template]) => [key, template.toObject()])
        ),
        rawTemplates: Object.fromEntries(templates.rawTemplates),
        templateKeys: Array.from(templates.templateKeys).sort((a, b) => {
            // Same ordering logic for template keys
            const isMainA = !a.includes('content');
            const isMainB = !b.includes('content');
            if (isMainA && !isMainB) return -1;
            if (!isMainA && isMainB) return 1;
            return a.localeCompare(b);
        })
    };
    return JSON.stringify(obj, null, indented ? 2 : 0).replace(/\r\n/g, '\n');
}

/**
 * Serializes preprocessed summary to JSON format
 * @param {Object} summary - The summary object to serialize
 * @param {boolean} indented - Whether to indent the JSON output
 * @returns {string} JSON string representation
 */
export function serializePreprocessedSummary(summary, indented = true) {
    return JSON.stringify(summary, null, indented ? 2 : 0).replace(/\r\n/g, '\n');
}

File: rust/AssemblerTest/src/preprocess_extensions.rs
// PreprocessExtensions: Methods for summary and JSON conversion
use assembler::template_model::model_preprocess::{PreprocessedSiteTemplates, PreprocessedSummary};

pub struct PreprocessJsonConverter;

impl PreprocessJsonConverter {
    /// Serializes a PreprocessedSiteTemplates to JSON format matching C# structure
    pub fn serialize_preprocessed_templates(templates: &PreprocessedSiteTemplates, indented: bool) -> String {
        if indented {
            Self::to_pretty_json(templates)
        } else {
            Self::to_json(templates)
        }
    }

    /// Serializes a PreprocessedSummary to JSON format matching C# structure
    pub fn serialize_preprocessed_summary(summary: &PreprocessedSummary, indented: bool) -> String {
        if indented {
            Self::to_pretty_json(summary)
        } else {
            Self::to_json(summary)
        }
    }

    fn to_json<T: serde::Serialize>(value: &T) -> String {
        serde_json::to_string(value).unwrap_or_default()
    }

    fn to_pretty_json<T: serde::Serialize>(value: &T) -> String {
        serde_json::to_string_pretty(value).unwrap_or_default()
    }
}

pub trait PreprocessExtensions {
    fn create_summary(&self) -> PreprocessedSummary;
    fn to_json(&self, indented: bool) -> String;
    fn to_summary_json(&self, indented: bool) -> String;
}

impl PreprocessExtensions for PreprocessedSiteTemplates {
    fn create_summary(&self) -> PreprocessedSummary {
        let templates_requiring_processing = self.templates.values()
            .filter(|t| t.requires_processing())
            .count();
        let templates_with_json_data = self.templates.values()
            .filter(|t| t.has_json_data())
            .count();
        let templates_with_placeholders = self.templates.values()
            .filter(|t| t.has_placeholders())
            .count();
        PreprocessedSummary {
            site_name: self.site_name.clone(),
            templates_requiring_processing,
            templates_with_json_data,
            templates_with_placeholders,
            total_templates: self.templates.len(),
        }
    }

    fn to_json(&self, indented: bool) -> String {
        PreprocessJsonConverter::serialize_preprocessed_templates(self, indented)
    }

    fn to_summary_json(&self, indented: bool) -> String {
        let summary = self.create_summary();
        PreprocessJsonConverter::serialize_preprocessed_summary(&summary, indented)
    }
}